RFC0003.md

```diff
@@ -0,0 +1,179 @@
+
+Some additional info keys are provided for the revised operations:
+
+* PMIX\_CONNECT\_NOTIFY\_EACH: generate a local event notification using the PMIX\_PROC\_HAS\_CONNECTED event each time a process connects, and the PMIX\_PROC\_HAS\_DISCONNECTED event each time a process disconnects. The PMIx server is responsible for generating these events.
+
+* PMIX\_CONNECT\_NOTIFY\_REQ: notify each of the indicated procs that they are requested to connect using the PMIX\_CONNECT\_REQUESTED event. The PMIx server is responsible for generating this event.
+
+* PMIX\_CONNECT\_OPTIONAL: participation is optional - do not return error if procs terminate without having connected. This is the joint responsibility of the host RM (to notify if remote procs terminate early) and the PMIx server.
+
+* PMIX\_CONNECT\_EARLY\_TERM\_ALLOWED: do not generate an error if a process terminates without disconnecting. The host RM and PMIx server are responsible for adjusting completion requirements on any ongoing or future collectives for this nspace.
+
+* PMIX\_CONNECT\_XCHG\_ONLY: ensure that all participating processes have a copy of the job-level info from each nspace, but do not assign a new nspace and rank
+
+
+#### PMIx\_Connect
+Record the specified processes as "connected". Both blocking and non-blocking
+versions are provided. This means that the resource manager should treat the
+failure of any process in the specified group as a reportable event, and take
+appropriate action. Note that different resource managers may respond to
+failures in different manners.
+
+The host RM will assign a unique nspace to the resulting process group. The
+new nspace must be provided when disconnecting from the group. All procs are
+required to disconnect from the nspace prior to terminating.
+
+Note: any single process _can_ be simultaneously engaged
+in multiple connect operations. For scalability, PMIx does not use a
+collective to assign an identifier to the connect operation. Instead, it
+is expected that the provided array of process IDs will serve to identify
+a specific PMIx\_Connect operation. This means, however, that there can
+only be _one_ active connect operation with that exact array of process
+IDs at a time. For these purposes, an array element consisting of an
+nspace and PMIX\_RANK\_WILDCARD counts as only a single element - i.e.,
+an array that specified every process individually for that nspace would
+result in a different signature than an array containing just the wildcard
+rank entry.
```

Comment:
Just to make sure I understand this `PMIX_RANK_WILDCARD` statement correctly.

In the eyes of connect, are these arrays valid, different requests (assume nspace `1` with 2 ranks):
`array1 = [{1, WILD}]`
`array2 = [{1, 0}, {1, WILD}]`
`array3 = [{1, 0}, {1, 1}]`

Reply:
My comment here is actually incorrect, at least per the current implementation. In the prototype branch, we check the array by proc, disregarding order. So your examples (both here an above) would all be considered the same request.

I can correct the comment, or we could decide to make it order-dependent. I'm inclined to leave it as order _independent_ as the alternative feels like we are stepping into specifying implementation.

Thoughts?

Reply:
I'm also inclined to keep it order independent. If we find that we need an order-preserving option then we could add an info key for it later - i don't think we need it right now.